Sorting existing languages in the menu

shurup opened the pull request:
### Describe your changes
As discussed before (see the issue below), ordering the localisations in the top-right menu seems a good idea. However, while implementing it, I realised that sorting localisations written in various languages can be quite tricky.

Based on [this ICU demo](https://icu4c-demos.unicode.org/icu-bin/locexp?_=en_US&d_=en&x=col) and some common sense (e.g. looking at how Google sorts the languages in its interfaces), I ended up with this order for existing languages:

- English (should be always first anyway)
- Deutsch
- Español (Spanish)
- Français (French)
- Italiano
- Português
- हिन्दी (Hindi)
- বাংলা (Bengali)
- 한국어 (Korean)  
- 简体中文 (Simplified Chinese)
- 繁體中文 (Traditional Chinese)
- اردو (Urdu)

I'll be happy to see any other thoughts on the best order (if there are any).

One last important thing to be aware of: we will have to update various `weight` values for various languages with each new localisation added (since this new language might have a higher position in the resulting menu).

### Related issue number or link (ex: `resolves #issue-number`)
Resolves #2480

### Checklist before opening this PR (put `x` in the checkboxes)
- [x] This PR does not contain plagiarism
  - don’t copy other people’s work unless you are quoting and contributing it to them.
- [x] I have signed off on all commits 
  - [signing off](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) (ex: `git commit -s`) is to affirm that commits comply [DCO](https://wiki.linuxfoundation.org/dco). If you are working locally, you could add an alias to your `gitconfig` by running `git config --global alias.ci "commit -s"`.

seokho-son reviewed (commented):
Hi @shurup 
Thanks for implementing https://github.com/cncf/glossary/issues/2480 :)

I agree with the current sorting method. (actually, I think the order depends on personal preferences)

Since you've sorted based on English words, it would be better to add (complete) English word for each language title to improve consistency.

languageName = "Português (Portuguese)"
languageName = "Italiano (Italian)"
languageName ="Deutsch (German)"

If German is acceptable, I guess we need to change the order for Deutsch

shurup commented:
> it would be better to add (complete) English word for each language title to improve consistency.

Great idea indeed, thank you!

> Since you've sorted based on English words

Just to clarify this moment: currently, the sorting is based not on the English words, but on some kind of transliteration of original language names into English. It means we consider _Español (Spanish)_ as a language name starting from _E_ (not from _S_). That's why _Deutsch_, which starts from _D_, seems to be in place — i.e., despite adding its English name after the original form, we still consider _Deutsch_ as the main word for sorting.

seokho-son commented:
Hi @shurup, my assumption about sorting by English was a misunderstanding. ;)

However, still, I believe having a clear and consistent criterion for ordering is preferable. From this perspective, I'd like to suggest listing based on the English words written in parentheses. This is important as we also need to update the guidelines accordingly.

In fact, as you mentioned, languages like Español and Deutsch can be read in English, so matching the sorting criteria to this makes sense. However, considering other languages and potential new additions, I want to avoid any special complications.

After all, local users will quickly find their language regardless of the order (for example, Koreans will easily spot `한국어`), and they are unlikely to search based on the English spelling in the parentheses. If the sorting is done in English, primarily for the benefit of English-speaking locals (mainly in the US), I don't see a major issue.

What do you think?

I'd like hear @nate-double-u and @iamNoah1 's opinion as well. :)

nate-double-u commented:
Do we need the English in parentheses? We can order it that way if we'd like, but i'm not sure we need the language name in English. I'm thinking about who the menu item is for -- English speakers will see "English" in English, so will be able to navigate if needed, and other language readers will see their language written in their language, so I'm not sure what the value of having each language's name listed in English would be.

I'm happy for the sorting to be done with this transliteration process, but we should update the https://github.com/cncf/glossary/blob/main/LOCALIZATION.md file with the new process.

iamNoah1 commented:
I agree with sorting based on english term, but also not sure if we need to add it in parentheses. But I have also no strong opinion about the later.

Okabe-Junya commented:
Hi there!!

FYI: In kubernetes/website repo as well, a similar PR has been created.

- https://github.com/kubernetes/website/pull/44869

If this is helpful for English speakers, I agree with merging this PR. (as son-san mentioned, many Japanese would look for "日本語" too).

Additionally, if we make this change, it seems more understandable to have config.toml in the order of weights. In other words, it seems easier to read if it is as follows:

```toml
[languages.foo]
...
 weight = 2

[languages.bar]
...
weight = 3
```

shurup commented:
Thank you everyone for your thoughts!

Personally, I think that following the English words to order languages **and** removing English in parentheses doesn't make much sense. Based on those criteria, we'll get such a list:

- বাংলা
- English _(or it might still go the first one)_
- Français
- Deutsch
- हिन्दी
- Italiano
- 한국어
- Português
- 简体中文
- Español
- 繁體中文
- اردو

… which seems to be quite random to me as a person who can't read any non-Latin symbols. (And I guess it will still be pretty awkward for those who can read more languages, but definitely not all of them.) Adding the English names in parentheses helps simply because I start seeing some logic behind this ordering :rofl: 

At the same time, perhaps the _transliteration_ of all languages is too difficult to explicitly explain and maintain (e.g., writing guidelines, etc.), which I can fully understand.

Thus, I think that sorting the English names and specifying those English names in parentheses is the best compromise. As I can see from the Kubernetes documentation (thanks, @Okabe-Junya, for sharing!), the same logic was applied there.

I also like the idea of reordering the language blocks in `config.toml` according to the newly determined UI-facing weights of languages.

Does this all sound reasonable to others? If so, I'm ready to make corresponding changes, including those in the "Initiating a new localization team" docs. Otherwise, let's decide which way will be better.

seokho-son commented:
Hi @nate-double-u @iamNoah1 @Okabe-Junya @shurup 

Thank you to everyone who participated in the discussion. (And a special thanks to @shurup for summing it up so well)
It seems we've reached a general consensus through our discussion. As I mentioned initially, given that the Glossary is viewed by audiences speaking various languages, it seemed unlikely that there would be a clear-cut sorting method. In open source, even if there's no single right answer, the best solution often emerges through discussion. :) Thanks again.

Here's a summary, and it would be great if @shurup could confirm and implement these:

- Add parentheses after the language names to include the Latin transcription (to be added for languages currently not having this).
- Sort each language based on the alphabetized transcription using the 'weight' in the config (but let's keep English first, as it's the source of the Glossary definitions).
- Arrange the order of language settings in the config also according to their weight.


Additional points for implementation:

- While resolving merge conflicts in PRs, add Japanese as a new language.
- Update the method for adding new language settings to the site configuration on https://github.com/cncf/glossary/blob/main/LOCALIZATION.md#3-1-adding-a-new-language-setting-to-site-configuration (currently, languages are added at the end).

shurup commented:
Thanks for this excellent summary (and your very detailed ToDo list), @seokho-son! I'll make relevant changes and let you know here.

shurup commented:
Hi @seokho-son! All changes are done. Please review them and let me know if something needs to be corrected.

Conflicts are also resolved now.

iamNoah1 commented on LOCALIZATION.md:
```diff
@@ -57,26 +57,31 @@ Open a PR with the localization initiation following this example: https://githu
-The `[languages]` block of `config.toml` is used to set the language. For instance, `[languages.en]` stands for English and `[languages.ko]` for Korean language configuration. Go to the `[languages]` block in `config.toml` and add a new block for your language-specific configuration. For instance, the Korean localization team added its `[languages.ko]` block after the `[languages.en]` block.
+The `[languages]` block of `config.toml` is used to set the language. For instance, `[languages.en]` stands for English and `[languages.ko]` for Korean language configuration. Go to the `[languages]` block in `config.toml` and find a place for your new localization:
+
+ - the English language always stays first;
```
```suggestion
 - the English language always stays first.
```

iamNoah1 commented on LOCALIZATION.md:
```diff
@@ -57,26 +57,31 @@ Open a PR with the localization initiation following this example: https://githu
+The `[languages]` block of `config.toml` is used to set the language. For instance, `[languages.en]` stands for English and `[languages.ko]` for Korean language configuration. Go to the `[languages]` block in `config.toml` and find a place for your new localization:
+
+ - the English language always stays first;
+ - other languages are sorted based on the alphabetized English translation of their names (see how currently existing localizations are ordered as an example);
```
```suggestion
 - other languages are sorted based on the alphabetized English translation of their names (see how currently existing localizations are ordered as an example).
```

iamNoah1 commented on LOCALIZATION.md:
```diff
@@ -57,26 +57,31 @@ Open a PR with the localization initiation following this example: https://githu
+
+ - the English language always stays first;
+ - other languages are sorted based on the alphabetized English translation of their names (see how currently existing localizations are ordered as an example);
+ - each localization is written in its own language followed by the English translation in parentheses — e.g., `한국어 (Korean)`;
```
```suggestion
 - each localization is written in its own language followed by the English translation in parentheses — e.g., `한국어 (Korean)`.
```

iamNoah1 reviewed (approved):
thanks @shurup LGTM

shurup commented:
Thanks, @iamNoah1! I applied your changes: dots instead of semicolons at the end of lists. My sense of beauty also suggested I should then capitalise the first letters in this list, which I did.

I also rebased this PR to `main` and fixed the order for the newly added Turkish localisation. If everything is good, it would be fantastic to merge it and avoid further rebasing/reordering for new languages :sweat_smile: :pray:

seokho-son reviewed (approved):
+1 approval to help @iamNoah1 make final decision. :)
